tensornetwork/contractors/stochastic_contractor.py

```diff
@@ -0,0 +1,102 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Stochastic Network Contraction."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import random
+import numpy as np
+from tensornetwork import tensornetwork
+
+
+def get_random_edge(network: tensornetwork.TensorNetwork) -> tensornetwork.Edge:
+  """Returns a random connected edge.
+  If it finds a trace edge it contracts it.
+  Args:
+    network: Connected TensorNetwork we are trying to contract.
+  Returns:
+    edge: A random connected edge from the network.
+  """
+  node = random.choice(tuple(network.nodes_set))
+  edge = random.choice(node.edges)
+  while edge.is_dangling():
+    edge = random.choice(node.edges)
+  if edge.node1 is edge.node2:
+    # Contract the trace edge and look for a new one
+    network.contract(edge)
+    return get_random_edge(network)
+  return edge
+
+
+def edge_cost(edge: tensornetwork.Edge) -> int:
+  """Calculates cost of contracting an edge.
+  If A and B are the tensors that share the given `edge`, cost is defined as:
+  cost = dims(A * B) - max(dims(A), dims(B)), where
+  * denotes contraction of all shared edges (`contract_between`) and
+  dims(X) is the total dimension of tensor X (product of sizes of all axes).
+  Args:
+    edge: Edge to calculate the cost.
+  Returns:
+    cost: Cost of the given edge.
+  """
+  # TODO: Verify whether this loss makes sense.
+  node1, node2 = edge.node1, edge.node2
+  # Calculate dimension of all shared edges
+  nodes = {node1, node2}
+  shared_dimension = 1
+  for edge in node1.edges:
+    if set(edge.get_nodes()) == nodes:
+      shared_dimension *= int(edge.node1.get_tensor().shape[edge.axis1])
+
+  dimension1 = int(np.prod(node1.get_tensor().shape))
+  dimension2 = int(np.prod(node2.get_tensor().shape))
+  prod_dimension = ((dimension1 // shared_dimension) *
+                    (dimension2 // shared_dimension))
+  cost = prod_dimension - max(dimension1, dimension2)
+  return cost
+
+
+def stochastic(network: tensornetwork.TensorNetwork,
+               max_rejections: int, threshold: int = -1
+              ) -> tensornetwork.TensorNetwork:
+  """Contracts a connected network by stochastically picking edges.
+  Algorithm 2 in page 7 of https://doi.org/10.1371/journal.pone.0208510.
+  Cost calculation is slightly modified here.
+  Args:
+    network: Connected TensorNetwork to contract fully.
+    max_rejections: Maximum number of rejections before you increase threshold.
+    threshold: Initial value for the threshold.
+  Returns:
+    network: TensorNetwork with a single node after fully contracting.
+  """
+  # TODO: Think about a proper threshold value for our loss.
+  network.check_connected()
+  rejections = 0
+
+  while len(network.nodes_set) > 1:
```

Comment:
This might cause some issues when the network is disconnected, and this also won't contract a single node with trace edges.

Reply:
This was designed to work only with connected networks (`network.check_connected` would raise an error for disconnected). In this case also the while loop in the first function would not break because it would never get a node with dangling edges only. Trace edges would still be a problem though.
Anyway, I changed it and now it should work with disconnected networks too (hopefully). The only drawback is the `random.choice(tuple(set))` which is probably suboptimal and has to run many times.